fix streaming

dtrifiro opened the pull request:
- implement http streaming, cleanup streaming implementations/docs

dtrifiro committed:
implement http streaming

dtrifiro commented:
Merge after #39

z103cb commented on examples/example.py:
```diff
@@ -3,21 +3,52 @@
 text = "What is the boilign point of Nitrogen?"
```
```suggestion
text = "What is the boiling point of Nitrogen?"
```

z103cb commented on tests/fixtures/mocked_results.py:
```diff
@@ -0,0 +1,113 @@
+    )
+
+
+# FIXME: Validate text stream mocking. There's a lot of logic here.
```
is this FIXME still applicable ?

z103cb reviewed (changes requested):
When executing pytest for this PR the MTLS, TLS tess for streaming are failing.

dtrifiro committed:
cleanup streaming APIs

dtrifiro committed:
refactor client interfaces

dtrifiro committed:
docs: update example

dtrifiro committed:
grpc client: add insecure flag

dtrifiro committed:
tests: disable stream tests for mocked server

See https://github.com/opendatahub-io/caikit-nlp-client/issues/46

dtrifiro committed:
http client: go back to the previous streaming implementation

This assumes the output format of caikit+tgis with the following image:

```
quay.io/opendatahub/caikit-tgis-servingcaikit-tgis-serving:fast-934c85f
```

Which looks like this:

```
data: {
data:   "generated_text": "",
data:   "tokens": [],
data:   "details": {
data:     "finish_reason": "NOT_FINISHED",
data:     "generated_tokens": 0,
data:     "seed": 0,
data:     "input_token_count": 8
data:   },
data:   "producer_id": null
data: }

data: {
...
data: }

...
```

dtrifiro committed:
http client: alternative streaming api impl

this implementation assumes that the streaming results look like this:

```
event: message
data: <json payload>

event: message
data: <json payload>
...
```

dtrifiro committed:
http client: silence requests mypy errors

dtrifiro commented on tests/test_http_client.py:
```diff
@@ -8,85 +11,75 @@ def test_generate_text(
+):
+    if not pytestconfig.option.real_caikit:
+        pytest.skip(
+            reason="stream mocking is broken, see https://github.com/opendatahub-io/caikit-nlp-client/issues/46"
```
I have no idea why this is broken, but it also makes tests hangs. See #46

z103cb commented on src/caikit_nlp_client/grpc_client.py:
```diff
@@ -159,32 +188,45 @@ def _close(self):
-                "A client key, client and server certificates have been detected, \
-                creating a mTLS channel"
+            log.info("Connecting using mTLS for secure channel")
+            # TODO: check if server_cert has the same meaning as ca_cert, and get
```
@dtrifiro, the code in the server is as follows:

```python
       if (
            self.tls_config
            and self.tls_config.server.key
            and self.tls_config.server.cert
        ):
            log.info("<RUN10001805I>", "Running with TLS")

            tls_server_pair = (
                bytes(self._load_secret(self.tls_config.server.key), "utf-8"),
                bytes(self._load_secret(self.tls_config.server.cert), "utf-8"),
            )
            if self.tls_config.client.cert:
                log.info("<RUN10001806I>", "Running with mutual TLS")
                # Combine the client cert with the server's own cert so that
                # health probes can use the server's key/cert instead of needing
                # one signed by a potentially-external CA.
                root_certificates = b"\n".join(
                    [
                        bytes(self._load_secret(self.tls_config.client.cert), "utf-8"),
                        tls_server_pair[1],
                    ]
                )
                # Client will verify the server using server cert and the server
                # will verify the client using client cert.
                server_credentials = grpc.ssl_server_credentials(
                    [tls_server_pair],
                    root_certificates=root_certificates,
                    require_client_auth=True,
                )
            else:
                server_credentials = grpc.ssl_server_credentials([tls_server_pair])
```

z103cb commented on src/caikit_nlp_client/grpc_client.py:
```diff
@@ -21,18 +22,36 @@ class GrpcClient:
+        ca_cert: Optional[bytes] = None,
+        client_key: Optional[bytes] = None,
+        client_cert: Optional[bytes] = None,
+        server_cert: Optional[bytes] = None,
```
The `server_cert` argument can be removed.

dtrifiro replied:
Done

z103cb commented on src/caikit_nlp_client/grpc_client.py:
```diff
@@ -21,18 +22,36 @@ class GrpcClient:
+            insecure=insecure,
+            client_key=client_key,
+            client_cert=client_cert,
+            server_cert=server_cert,
```
The server_cer can be removed

dtrifiro replied:
Done

z103cb commented on src/caikit_nlp_client/grpc_client.py:
```diff
@@ -159,32 +188,45 @@ def _close(self):
+            return grpc.insecure_channel(connection)
+
+        credentials_kwargs: dict[str, bytes] = {}
         if ca_cert:
```
With this logic (which can be simplified) using the only the CA certificate worked for me:

```python
        if ca_cert and not any((client_cert, client_key)):
            log.info("Connecting using provided CA certificate for secure channel")
            credentials_kwargs.update(root_certificates=ca_cert)
        elif client_key and client_cert and ca_cert:
            log.info("Connecting using mTLS for secure channel")
            # TODO: check if server_cert has the same meaning as ca_cert, and get
            # rid of it if so
            credentials_kwargs.update(
                root_certificates=ca_cert,
                private_key=client_key,
                certificate_chain=client_cert,
            )
        elif any((client_cert, client_key, ca_cert)):
            raise ValueError("mTLS requires client_cert, client_key and ca_cert")
```

dtrifiro replied:
Done

z103cb reviewed (commented):
I have validated the mTLS connection would work with only ca certificates. Some of the TODO(s) can be removed.

dtrifiro committed:
tests: cleanup, don't validate responses contents

dtrifiro committed:
grpc client: replace server_cert with ca_cert
